sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala

```diff
@@ -606,7 +606,7 @@ case class NTile(buckets: Expression) extends RowNumberLike with SizeBasedWindow
     zero,
     zero,
     zero,
-    (n / buckets).cast(IntegerType),
+    (n.cast(DecimalType.IntDecimal) / buckets.cast(DecimalType.IntDecimal)).cast(IntegerType),
```

Comment:
`DeclarativeAggregate` (e.g., `updateExpressions`) uses DSLs `org.apache.spark.sql.catalyst.dsl._` in many places. IIUC these DSLs are assumed to be checked by the analyzer (e.g., type coercion), but some DSLs in `DeclarativeAggregate` are evaluated directly without the analyzer check. I personally think this is error-prone...

Reply:
the DSL is just a shortcut of writing expressions directly. It's error-prone to create expressions that won't go through analyzer, but it's nothing to do with DSL.